pkg/plugins/mellanox/mellanox_plugin.go

```diff
@@ -115,29 +115,47 @@ func (p *MellanoxPlugin) OnNodeStateChange(old, new *sriovnetworkv1.SriovNetwork
 		if err != nil {
 			return false, false, err
 		}
-		if fwCurrent.enableSriov != fwNext.enableSriov || fwCurrent.totalVfs != fwNext.totalVfs {
-			needReboot = true
-			continue
-		}
 
 		isDualPort := isDualPort(ifaceSpec.PciAddress)
 		// Attributes to change
 		attrs := &mlnxNic{totalVfs: -1}
+		var changeWithoutReboot bool
 
 		var totalVfs int
-		totalVfs, needReboot = handleTotalVfs(fwCurrent, attrs, ifaceSpec, isDualPort)
+		totalVfs, needReboot, changeWithoutReboot = handleTotalVfs(fwCurrent, fwNext, attrs, ifaceSpec, isDualPort)
 		needReboot = handleEnableSriov(totalVfs, fwCurrent, attrs) || needReboot
 		needLinkChange, err := handleLinkType(pciPrefix, fwCurrent, attrs)
 		if err != nil {
 			return false, false, err
 		}
 
 		needReboot = needReboot || needLinkChange
-		if needReboot {
+		if needReboot || changeWithoutReboot {
 			attributesToChange[ifaceSpec.PciAddress] = *attrs
 		}
 	}
 
+	// Set total VFs to 0 for mellanox interfaces with no spec
+	for pciPrefix := range mellanoxNicsStatus {
+		if _, ok := processedNics[pciPrefix]; ok {
+			continue
+		}
+
+		// Add the nic to processed Nics to not repeat the process for dual nic ports
+		processedNics[pciPrefix] = true
+		pciAddress := pciPrefix + "0"
+
+		_, fwNext, err := getMlnxNicFwData(pciAddress)
+		if err != nil {
+			return false, false, err
+		}
+
+		if fwNext.totalVfs > 0 || fwNext.enableSriov {
```

Comment:
@Mmduh-483  nit: if fwNext.enableSriov is disabled, will it still be possible that fwNext.totalVfs is larger than 0?
My concern is when `fwNext.enableSriov` is used, it may try to configure devices that are not supported (for example, old CX-3 device which has enableSriov configured to true by default). and we don't guarantee this will work with mstconfig or any openshift platform.

Reply:
Updated it to skip unsupported devices

Reply:
@Mmduh-483 I just tried this combined settings `SRIOV_EN=false` && `NUM_OF_VFS=4` and found that they can be configured successfully to the firmware.  Is this expected? my understand was if SR-IOV_ENV is disabled, it's not possible to configure num_vfs.

For the fix, I was thinking we can simply remove the `fwNext.enableSriov` because if `fwNext.totalVfs > 0` is true, it means operator has successfully configured the device (which should be a supported device).

Reply:
@zshi-redhat yes it is possible  to set `SRIOV_EN` to false and set the `NUM_OF_VFS` (total vfs), when set `SRIOV_EN` to false, the files `/sys/class/net/$netdev/device/sriov_numvfs` and `sriov_totalvfs` disappeared, so you can't create vfs.

Reply:
As a note:
`SRIOV_EN` configures the NIC to whether or not expose SR-IOV capability to the host
`NUM_OF_VFS` set the maximal number of VFS the NIC supports

The parameters dont depend on each other. which is a good thing or else you would have needed to perform two operations and possibly an additional reboot if they were.

Reply:
@Mmduh-483  @adrianchiris  thanks for the note and explaination!

Reply:
Sure :)